librz/analysis/var_global.c

```diff
@@ -0,0 +1,335 @@
+	}
+	if (!rz_rbtree_aug_insert(&analysis->global_var_tree, &global_var->addr, &global_var->rb, global_var_node_cmp, NULL, NULL)) {
+		return false;
+	}
+	return true;
+}
+
+/**
+ * \brief Free the global variable instance
+ * 
+ * \param glob Global variable instance
+ * \return void
+ */
+RZ_API void rz_analysis_var_global_free(RzAnalysisVarGlobal *glob) {
+	if (!glob) {
+		return;
+	}
+
+	RZ_FREE(glob->name);
+	rz_type_free(glob->type);
+	rz_vector_fini(&glob->constraints);
+	RZ_FREE(glob);
+}
+
+/**
+ * \brief Delete and free the global variable by its name
+ * 
+ * \param analysis RzAnalysis
+ * \param name Global Variable name
+ * \return true if succeed
+ */
+RZ_API bool rz_analysis_var_global_delete_byname(RzAnalysis *analysis, RZ_NONNULL const char *name) {
+	rz_return_val_if_fail(analysis && name, false);
+	RzAnalysisVarGlobal *glob = rz_analysis_var_global_get_byname(analysis, name);
+	if (!glob) {
+		return false;
+	}
+	// We need to delete RBTree first because ht_pp_delete will free its member
+	bool deleted = rz_rbtree_delete(&analysis->global_var_tree, &glob->addr, global_var_node_cmp, NULL, NULL, NULL);
+	return deleted ? ht_pp_delete(analysis->ht_global_var, name) : deleted;
+}
+
+/**
+ * \brief Same as rz_analysis_var_global_delete_byname but by its address
+ * 
+ * \param analysis RzAnalysis
+ * \param addr Global Variable address
+ * \return true if succeed
+ */
+RZ_API bool rz_analysis_var_global_delete_byaddr(RzAnalysis *analysis, ut64 addr) {
+	rz_return_val_if_fail(analysis, false);
+	RzAnalysisVarGlobal *glob = rz_analysis_var_global_get_byaddr(analysis, addr);
+	if (!glob) {
+		return false;
+	}
+	// We need to delete RBTree first because ht_pp_delete will free its member
+	bool deleted = rz_rbtree_delete(&analysis->global_var_tree, &glob->addr, global_var_node_cmp, NULL, NULL, NULL);
+	return deleted ? ht_pp_delete(analysis->ht_global_var, glob->name) : deleted;
+}
+
+/**
+ * \brief Get the instance of global variable by its name
+ * 
+ * \param analysis RzAnalysis
+ * \param name Global variable name
+ * \return RzAnalysisVarGlobal *
+ */
+RZ_API RZ_BORROW RzAnalysisVarGlobal *rz_analysis_var_global_get_byname(RzAnalysis *analysis, RZ_NONNULL const char *name) {
+	rz_return_val_if_fail(analysis && name, NULL);
+	return (RzAnalysisVarGlobal *)ht_pp_find(analysis->ht_global_var, name, NULL);
+}
+
+struct list_addr {
+	RzList /* <RzAnalysisVarGlobal> */ *list;
+	ut64 addr;
+};
+
+/**
+ * \brief Same as rz_analysis_var_global_get_byname but by its address
+ * 
+ * \param analysis RzAnalysis
+ * \param addr Global variable address
+ * \return RzAnalysisVarGlobal *
+ */
+RZ_API RZ_BORROW RzAnalysisVarGlobal *rz_analysis_var_global_get_byaddr(RzAnalysis *analysis, ut64 addr) {
+	rz_return_val_if_fail(analysis, NULL);
+	RBIter it;
+	RzAnalysisVarGlobal *node, *tmp = NULL;
+	rz_rbtree_foreach (analysis->global_var_tree, it, node, RzAnalysisVarGlobal, rb) {
```

Comment:
I think this should not be rz_rbtree_foreach, right? Otherwise you lose all the advantage of using an Rbtree if you always have to iterate over all elements. something like rz_rbtree_lower_bound_forward should probably be used. cc @thestr4ng3r

Reply:
It would be `rz_rbtree_upper_bound()` here. Also it is necessary to disambiguate what "by its address" means. Usually we have two functions:
* `rz_<...>_get_<x>_at` which gives you the thing that starts exactly at the given address. (use `rz_rbtree_find()`)
* `rz_<...>_get_<x>_in` which gives you the thing that contains the given address. (use `rz_rbtree_upper_bound()` + a single check against the size of the type)

I would recommend to do this here too.